File: backend/src/main/java/com/backend/application/service/ProductVariantOptionService.java
package com.backend.application.service;

import java.util.List;

import com.backend.application.dto.request.CreateProductVariantOptionRequest;
import com.backend.application.dto.request.UpdateProductVariantOptionRequest;
import com.backend.application.dto.response.ProductVariantOptionResponse;

public interface ProductVariantOptionService {

    List<ProductVariantOptionResponse> findAll();

    ProductVariantOptionResponse findById(Long id);

    ProductVariantOptionResponse create(CreateProductVariantOptionRequest request);

    ProductVariantOptionResponse update(Long id, UpdateProductVariantOptionRequest request);

    void delete(Long id);
}

File: storefront/src/app/modules/admin/custom/products/services/product-variant-option.service.ts
import { Injectable } from '@angular/core';
import { CrudEndpoints, CrudHttpService } from '@core/crud/crud-http.service';
import { Observable } from 'rxjs';
import {
    CreateProductVariantOptionRequest,
    ProductVariantOption,
    UpdateProductVariantOptionRequest,
} from '../models/product-variant-option.types';

@Injectable({ providedIn: 'root' })
export class ProductVariantOptionService extends CrudHttpService<
    ProductVariantOption,
    CreateProductVariantOptionRequest,
    UpdateProductVariantOptionRequest
> {
    protected override endpoints: CrudEndpoints = {
        list: 'productVariantOptions',
        getById: 'productVariantOptionById',
        create: 'productVariantOptions',
        update: 'productVariantOptionById',
        delete: 'productVariantOptionById',
    };

    getAll(): Observable<ProductVariantOption[]> {
        return this.customGet<ProductVariantOption[]>(
            'productVariantOptions',
            {}
        );
    }
}
